### PR TITLE
Add a password expiration config option (resolves #2485)

### LINKED ISSUE #2485: Track user last password change date and add an option to force password changes
We would like to add a config option, "password_expiration_period" which would force users to change their password every so many days. Users will be redirected to the change password screen when they log in after their password expires. If the config option does not exist or is 0, behavior would be the same as it is currently. If no objection, I'll make a pull request shortly.

### PR DESCRIPTION
**Pull request by benfulton**
Add a "last_password_change" field to user table and user object.
Add a config option "password expiration_period" to allow admins
to require users to change passwords every so many days. On login,
check for the presence of the config option and if the user pw
has expired, redirect to the "change password" page.

### CODE REVIEW

File: lib/galaxy/webapps/galaxy/controllers/user.py [hunk 1]
```diff
@@ -563,6 +563,19 @@ def __validate_login( self, trans, **kwd ):
                 message, status = self.resend_verification_email( trans, user.email, user.username )
         else:  # activation is OFF
             message, success, status = self.proceed_login( trans, user, redirect )
+            pw_expires = trans.app.config.password_expiration_period
+            if pw_expires:
+                if user.last_password_change < datetime.today() - pw_expires:
+                    trans.response.send_redirect(web.url_for(controller='user',
+                                                             action='change_password',
+                                                             message='Your time is up! Change your password to access galaxy',
```
Comment (martenson) [inline lib/galaxy/webapps/galaxy/controllers/user.py]: Please rephrase. Maybe something like `Your password has expired. Please change it to access Galaxy.`

**Comment by hexylena** [comment]
:+1: on the idea (any implementation issues aside). This will be nice to have!

File: lib/galaxy/model/migrate/versions/0132_add_lastpasswordchange_to_user.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+"""
+Migration script to add a last_password_change field to
```
Comment (martenson) [inline lib/galaxy/model/migrate/versions/0132_add_lastpasswordchange_to_user.py]: to the User table?

File: lib/galaxy/config.py [hunk 3]
```diff
@@ -113,6 +113,7 @@ def __init__( self, **kwargs ):
 
         self.expose_user_name = kwargs.get( "expose_user_name", False )
         self.expose_user_email = kwargs.get( "expose_user_email", False )
+        self.password_expiration_period = timedelta(days=int(kwargs.get( "password_expiration_period", 0 )))
```
Comment (martenson) [inline lib/galaxy/config.py]: please adhere to the codestyle of the file and add space before and after a bracket content

**Comment by martenson** [comment]
Looks very nice, added some minor comments inline, will test in action later today. Thank you @benfulton !

**Comment by martenson** [comment]
Thank you for the changes. I might get to testing it after the weekend, sorry.
@galaxybot test this

**Comment by martenson** [comment]
Tool Shed inherits the user controller but has separate database, we probably want to extend the functionality of expiring password to it. I can have a stab.

**Comment by nsoranzo** [comment]
This is changing the user permissions on 2 files: lib/galaxy/webapps/galaxy/controllers/user.py and lib/galaxy/webapps/galaxy/controllers/user.py

This should be reverted.

**Comment by martenson** [comment]
@nsoranzo good catch, I didn't even know github shows these changes

**Comment by dannon** [comment]
https://github.com/benfulton/galaxy/pull/1 fixes file permissions and has one more change inc addressing incorrect login logic

**Comment by dannon** [comment]
@galaxybot test this

**Comment by hexylena** [comment]
@galaxybot test this

(from request on IRC)

**Comment by dannon** [comment]
@benfulton Thanks for following up on the toolshed config!

(failing test is travis osx fail, not code fail)

**Pull request merged**